mlir/lib/Quantum/Transforms/adjoint_lowering.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "adjoint"
+
+#include <memory>
+#include <vector>
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Index/IR/IndexDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace llvm;
+using namespace mlir;
+
+namespace catalyst {
+namespace quantum {
+
+struct AdjointLoweringPass : public OperationPass<ModuleOp> {
+    AdjointLoweringPass() : OperationPass<ModuleOp>(TypeID::get<AdjointLoweringPass>()) {}
+    AdjointLoweringPass(const AdjointLoweringPass &other) : OperationPass<ModuleOp>(other) {}
+
+    StringRef getName() const override { return "AdjointLoweringPass"; }
+
+    StringRef getArgument() const override { return "adjoint-lowering"; }
+
+    StringRef getDescription() const override
+    {
+        return "Lower adjoint regions containing a single quantum operations.";
+    }
+
+    void getDependentDialects(DialectRegistry &registry) const override
+    {
+        registry.insert<arith::ArithDialect>();
+        registry.insert<linalg::LinalgDialect>();
+        registry.insert<index::IndexDialect>();
+        registry.insert<tensor::TensorDialect>();
+        registry.insert<memref::MemRefDialect>();
+        registry.insert<bufferization::BufferizationDialect>();
```

Comment:
I'm surprised, does the pass generate any ops that aren't already present in the input?

Reply:
Hmm, but do we need to describe the set of output ops here? This pass can definitely produce any of those as output by copying it from input.